File: 844-BackspaceStringCompare/844-BackspaceStringCompare.java
/*

844. Backspace String Compare
Easy

Given two strings S and T, return if they are equal when both are typed into empty text editors. # means a backspace character.

Example 1:

Input: S = "ab#c", T = "ad#c"
Output: true
Explanation: Both S and T become "ac".

Example 2:

Input: S = "ab##", T = "c#d#"
Output: true
Explanation: Both S and T become "".

Example 3:

Input: S = "a##c", T = "#a#c"
Output: true
Explanation: Both S and T become "c".

Example 4:

Input: S = "a#c", T = "b"
Output: false
Explanation: S becomes "c" while T becomes "b".

Note:

    1 <= S.length <= 200
    1 <= T.length <= 200
    S and T only contain lowercase letters and '#' characters.

Follow up:

    Can you solve it in O(N) time and O(1) space?

*/

class Solution {
    public boolean backspaceCompare(String S, String T) {
        Stack<Character> sStack = new Stack<>();
        Stack<Character> tStack = new Stack<>();
        for (int i = 0; i < S.length(); i++){
            if(S.charAt(i) == '#'){
                if (!sStack.isEmpty())
                    sStack.pop();
                continue;
            }
              sStack.push(S.charAt(i));
        }
        for (int i = 0; i < T.length(); i++){
            if(T.charAt(i) == '#'){
                if (!tStack.isEmpty())
                    tStack.pop();
                continue;
            }
            tStack.push(T.charAt(i));
        }
        while(!sStack.isEmpty() && !tStack.isEmpty()){
            if(sStack.pop() != tStack.pop())
                return false;
        }
        if (sStack.isEmpty() && tStack.isEmpty())
            return true;
        return false;
    }
}

class Solution {
    public boolean backspaceCompare(String s, String t) {
        String s1 = getString(s);
        String s2 = getString(t);
        return s1.equals(s2);
    }
    private String getString(String s){
        StringBuilder sb = new StringBuilder();
        for(char ch : s.toCharArray()){
            if(ch == '#' && sb.length() > 0) sb.deleteCharAt(sb.length() - 1);
            else if(ch != '#') sb.append(ch);
        }
        return sb.toString();
    }
}

File: 844-BackspaceStringCompare/844-BackspaceStringCompare.py
"""

844. Backspace String Compare

Given two strings s and t, return true if they are equal when both are typed into empty text editors. '#' means a backspace character.

Note that after backspacing an empty text, the text will continue empty.

Example 1:

Input: s = "ab#c", t = "ad#c"
Output: true
Explanation: Both s and t become "ac".

Example 2:

Input: s = "ab##", t = "c#d#"
Output: true
Explanation: Both s and t become "".

Example 3:

Input: s = "a#c", t = "b"
Output: false
Explanation: s becomes "c" while t becomes "b".

Constraints:

    1 <= s.length, t.length <= 200
    s and t only contain lowercase letters and '#' characters.

 

Follow up: Can you solve it in O(n) time and O(1) space?

"""

class Solution:
    def backspaceCompare(self, s: str, t: str) -> bool:
        s1, s2 = [], []
        def fill_stack(stack, string):
            for ch in string:
                if ch == "#":
                    if not stack: continue
                    stack.pop()
                else: stack.append(ch)
            return stack
        s1 = fill_stack(s1, s)
        s2 = fill_stack(s2, t)
        return True if s1 == s2 else False
